### PR TITLE
fix: Fix blocked and oom issues for scan lance

### PR DESCRIPTION
**Pull request by caican00**
## Changes Made

When building Scan Tasks, avoid directly binding the lance dataset object to each Scan Task. And the size of a single lance dataset is approximately 0.76MB.
`self._ds: 0.76 MB`

First, split the attributes of the lance dataset. The Scan Task only binds the attributes of the lance dataset.  When it is actually executed, refactor it back to the lance dataset to reduce the size of the execution plan.

For such an execution plan containing 8,354 scantasks, the size of the optimized execution plan has been reduced from over 13GB to 25MB.

The execution time ranges from execution failure to returning results within seconds.

```
== Physical Plan ==

* Limit: 10
|
* ScanTaskSource:
**|   Num Scan Tasks = 8354**
|   Estimated Scan Bytes = 0
|   Pushdowns: {limit: 10}
|   Schema: {seq_id#Int64, ext#Utf8, file_size_in_byte#Int64, program_lang#Utf8,
|     doc_type#Utf8, text#Utf8, filename#Utf8, repo_name#Utf8, sub_path#Utf8,
|     stars#Utf8, dataset#Utf8, pt#Utf8, tar_path#Utf8, file_path#Utf8}
|   Scan Tasks: [
|   {daft.io.lance.lance_scan:_lancedb_table_factory_function}
|   {daft.io.lance.lance_scan:_lancedb_table_factory_function}
|   {daft.io.lance.lance_scan:_lancedb_table_factory_function}
|   ...
|   {daft.io.lance.lance_scan:_lancedb_table_factory_function}
|   {daft.io.lance.lance_scan:_lancedb_table_factory_function}
|   {daft.io.lance.lance_scan:_lancedb_table_factory_function}
|   ]

```

## Related Issues

https://github.com/Eventual-Inc/Daft/issues/5591

## How to test
existing UTs.

## Checklist

- [ ] Documented in API Docs (if applicable)
- [ ] Documented in User Guide (if applicable)
- [ ] If adding a new documentation page, doc is added to `docs/mkdocs.yml` navigation
- [ ] Documentation builds and is formatted properly

### CODE REVIEW

File: daft/io/lance/lance_scan.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 from daft.recordbatch import RecordBatch
 
 from ..pushdowns import SupportsPushdownFilters
+import lance
 
 if TYPE_CHECKING:
     import lance
```
[the remark this answers is not in the document]
Comment (caican00) [reply daft/io/lance/lance_scan.py]: fixed.

File: daft/io/lance/_lance.py [hunk 2]
```diff
@@ -124,6 +124,17 @@ def read_lance(
         default_scan_options=default_scan_options,
         metadata_cache_size_bytes=metadata_cache_size_bytes,
     )
+
+    ds._daft_open_kwargs = {
+        "storage_options": storage_options,
+        "version": version,
```
Comment (Jay-ju) [inline daft/io/lance/_lance.py:130]: I don't find here which parameter affects the OOM of ray head?

Comment (caican00) [reply daft/io/lance/_lance.py:130]:
```
    def stream_plan(
        self,
        plan: DistributedPhysicalPlan,
        partition_sets: dict[str, PartitionSet[ray.ObjectRef]],
    ) -> Iterator[RayMaterializedResult]:
        plan_id = plan.id()

        t=time.time();
        sz=len(cp.dumps(plan));
        print("plan pickle bytes:", sz, "secs:", time.time()-t)
        t1=time.time();
        sz1=len(cp.dumps(partition_sets));
        print("psets pickle bytes:", sz1, "secs:", time.time()-t1)

        ray.get(self.runner.run_plan.remote(plan, partition_sets))
        while True:
            materialized_result = ray.get(self.runner.get_next_partition.remote(plan_id))
            if materialized_result is None:
                break
            yield materialized_result
```

I added some logs for my test, printed out the size of the plan and the size of a single lance dataset, and found that the size of a single lance dataset reached 0.76MB.

And before optimization, the size of my `DistributedPhysicalPlan` exceeded 13G, and serialization was very slow.  After optimization, it was only 25M.

cc @Jay-ju

Comment (caican00) [reply daft/io/lance/_lance.py:130]: The lance dataset object itself is relatively large.

Comment (caican00) [reply daft/io/lance/_lance.py:130]: I will add some screenshots.

Comment (caican00) [reply daft/io/lance/_lance.py:130]: > I will add some screenshots.

It seems that screenshots cannot be uploaded. I'll provide the test print logs.

The size of a single lance dataset:
```
self._ds: 0.76 MB
schema: 0.00 MB
```

Before optimization:
```
plan pickle bytes: 13298686802 bytes, secs: 596.3822300434113
psets pickle bytes: 5, secs: 4.6253204345703125e-05
```

Optimized
```
plan pickle bytes: 25834890 bytes, secs: 0.7463228702545166
psets pickle bytes: 5, secs: 1.8596649169921875e-05
```

**Comment by caican00** [comment]
Hi @colin-ho could you help review this pr? thanks!

File: daft/io/lance/_lance.py (hunk 2 above)
Comment (caican00) [reply daft/io/lance/_lance.py:130]: gently ping @Jay-ju could you help review it again? thank you!

**Comment by caican00** [comment]
gently ping @jaychia may i ask could you help review this pr? Thank you!

**Comment by caican00** [comment]
> ### Greptile Summary
> * Optimizes Lance scan task serialization by storing dataset URI and open kwargs instead of the full dataset object in each scan task
> * Adds `_daft_open_kwargs` attribute to dataset and modifies factory functions to reconstruct dataset from URI during execution
> * Includes comprehensive test coverage for serialization size verification across various fragment counts and edge cases
> 
> ### Confidence Score: 0/5
> * This PR has critical bugs that completely defeat its optimization purpose and will cause runtime errors
> * Two critical bugs on lines 208 and 238 pass the full `LanceDataset` object instead of `ds.uri` and `open_kwargs` to factory functions, completely defeating the optimization and likely causing type errors at runtime since the factory expects a string URI
> * Pay critical attention to `daft/io/lance/lance_scan.py` lines 208 and 238 - these bugs must be fixed before merging
> 
> ### Important Files Changed
> Filename	Overview
> daft/io/lance/lance_scan.py	Modified factory functions to accept URI+kwargs instead of dataset object, but has critical bugs on lines 208 and 238 that pass the full dataset object, defeating the optimization
> daft/io/lance/_lance.py	Added `_daft_open_kwargs` attribute to lance dataset to store reconstruction parameters for optimization
> tests/io/lancedb/test_lance_scan_task_optimization.py	New test file with comprehensive serialization size tests, but doesn't test limit pushdown path which has critical bugs
> ### Sequence Diagram
> ```mermaid
> sequenceDiagram
>     participant User
>     participant read_lance
>     participant LanceDataset
>     participant LanceDBScanOperator
>     participant ScanTask
>     participant Worker
>     participant factory_function
>     
>     User->>read_lance: read_lance(url, options)
>     read_lance->>LanceDataset: lance.dataset(url, **options)
>     LanceDataset-->>read_lance: dataset object
>     read_lance->>LanceDataset: attach _daft_open_kwargs
>     read_lance->>LanceDBScanOperator: create operator(ds)
>     LanceDBScanOperator->>ScanTask: to_scan_tasks()
>     Note over ScanTask: Creates lightweight tasks with<br/>ds.uri + open_kwargs instead<br/>of full dataset object
>     ScanTask->>Worker: serialize and send task
>     Worker->>factory_function: execute(ds_uri, open_kwargs, fragments)
>     factory_function->>LanceDataset: lance.dataset(ds_uri, **open_kwargs)
>     LanceDataset-->>factory_function: reconstructed dataset
>     factory_function->>factory_function: scan fragments
>     factory_function-->>Worker: record batches
>     Worker-->>User: results
> ```

fixed.

**Comment by colin-ho** [comment]
@greptileai

File: daft/io/lance/_lance.py [hunk 3]
```diff
@@ -124,6 +124,17 @@ def read_lance(
...
+        "asof": asof,
+        "block_size": block_size,
+        "commit_lock": commit_lock,
+        "index_cache_size": index_cache_size,
+        "default_scan_options": default_scan_options,
+        "metadata_cache_size_bytes": metadata_cache_size_bytes,
+    }
```
[the remark this answers is not in the document]
Comment (caican00) [reply daft/io/lance/_lance.py]: Thank you for you review @colin-ho 

https://github.com/lance-format/lance/blob/f95d081a3c8b54efe36cd1c47d8f56a6c218f60f/python/python/lance/dataset.py#L460-L528

yes, the serialization and serialization of `LanceDataset` do not include the `commit_lock` object, so here we can remove it to achieve the same effect as directly passing the `LanceDataset` object before. WDYT?

[the remark this answers is not in the document]
Comment (caican00) [reply daft/io/lance/_lance.py]: updated.

**Comment by caican00** [comment]
fixed UTs, hi @colin-ho do you have time to review again? thank you

File: tests/io/lancedb/test_lancedb_factory_function.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+                )
+        assert "Unable to import the `lance` package" in str(ei.value)
+        assert "pip install daft[lance]" in str(ei.value)
+    finally:
+        if saved is not None:
+            sys.modules["lance"] = saved
+
+
+@pytest.mark.parametrize("num_frags", [1, 2, 4])
+@pytest.mark.parametrize("use_filter", [False, True])
```
Comment (colin-ho) [inline tests/io/lancedb/test_lancedb_factory_function.py]: Put the arrow filter directly in the parameter so that you don't need to put an if block in the test itself.

Same for the required_cols

File: tests/io/lancedb/test_lancedb_factory_function.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+@pytest.mark.parametrize("use_required_cols", [False, True])
+@pytest.mark.parametrize("push_limit", [None, 3])
+def test_reconstructs_dataset_and_reads_fragments(
+    tmp_path_factory, num_frags, use_filter, use_required_cols, push_limit
+):
+    lance = pytest.importorskip("lance")
```
Comment (colin-ho) [inline tests/io/lancedb/test_lancedb_factory_function.py]: Just put this at the top of the module

File: tests/io/lancedb/test_lancedb_factory_function.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+            for k in agg.keys():
+                agg[k].extend(pdict[k])
+
+    assert agg is not None
+
+    expect_tbl = pa.Table.from_batches(direct_batches)
+    expect_pydict = expect_tbl.to_pydict()
+
+    # Order is deterministic across both since fragments and scanner settings match
+    assert agg == expect_pydict
```
Comment (colin-ho) [inline tests/io/lancedb/test_lancedb_factory_function.py]: I think you can replace all of this with just
```
for out_rb, direct_rb in zip(out_batches, direct_batches):
        assert out_rb.to_arrow_record_batch().to_pydict() == direct_rb.to_pydict()
```

File: tests/io/lancedb/test_lancedb_factory_function.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+        else:
+            for k in agg.keys():
+                agg[k].extend(pdict[k])
+
+    assert agg is not None
+
+    expect_tbl = pa.Table.from_batches(direct_batches_v1)
+    expect_pydict = expect_tbl.to_pydict()
+
+    assert agg == expect_pydict
```
Comment (colin-ho) [inline tests/io/lancedb/test_lancedb_factory_function.py]: Same here

**Comment by colin-ho** [comment]
@greptileai

File: tests/io/lancedb/test_lancedb_factory_function.py (hunk 4 above)
Comment (caican00) [reply tests/io/lancedb/test_lancedb_factory_function.py]: got it, let me fix it.

File: tests/io/lancedb/test_lancedb_factory_function.py (hunk 7 above)
Comment (caican00) [reply tests/io/lancedb/test_lancedb_factory_function.py]: updated

File: tests/io/lancedb/test_lancedb_factory_function.py (hunk 6 above)
Comment (caican00) [reply tests/io/lancedb/test_lancedb_factory_function.py]: updated.

File: tests/io/lancedb/test_lancedb_factory_function.py (hunk 5 above)
Comment (caican00) [reply tests/io/lancedb/test_lancedb_factory_function.py]: updated.

File: tests/io/lancedb/test_lancedb_factory_function.py (hunk 4 above)
Comment (caican00) [reply tests/io/lancedb/test_lancedb_factory_function.py]: updated.

**Comment by caican00** [comment]
Hi @colin-ho all comments have been addressed and could you help review again? Thank you very much.
